Python/crossinterp.c

```diff
@@ -713,6 +713,76 @@ _bool_shared(PyThreadState *tstate, PyObject *obj,
     return 0;
 }
 
+struct _shared_tuple_data {
+    Py_ssize_t len;
+    _PyCrossInterpreterData ** data;
```

Comment:
FWIW, I was going to suggest allocating this array and each item all at once:

```suggestion
    _PyCrossInterpreterData *data[0];
```

```c
static int
_tuple_shared(PyThreadState *tstate, PyObject *obj,
             _PyCrossInterpreterData *data)
{
    Py_ssize_t len = PyTuple_GET_SIZE(obj);
    if (len < 0) {
        return -1;
    }
    const size_t size = (sizeof(struct _shared_tuple_data)
                                   + (sizeof(_PyCrossInterpreterData) * len));
    if (_PyCrossInterpreterData_InitWithSize(
            data, tstate->interp, size, obj, _new_tuple_object
            ) < 0)
    ...
```

Then you don't have to allocate/free the array separately, nor each item.  You would, of course, need to change `shared->data[i]` to `&shared->data[i]` in all the relevant places.

Note that this hypothetically makes `_tuple_shared_free()` a little trickier.  On the one hand, it means you don't have to free each item, nor the array.  On the other hand, you have to deal with the (remote) possibility that the tuple's cross-interpreter data might be tied to a different interpreter than one or more of the items.  If that were to happen, the `_PyCrossInterpreterData_Release()` call in `_tuple_shared_free()` would cause the item to be released "later" in the other interpreter via a pending call.  There's a race between when that happens and when the tuple's cross-interpreter data (`data->data` here) is freed.  If the latter happens first then you get a crash.

As long as we know that all items belong to the same interpreter as the tuple then we're fine, and we can be confident about that (and throw in an assert or two to verify).

I said "was going to" because, ultimately I don't think allocating the parts separately or all at once makes a huge difference.  That said, you're welcome to take the all-at-once approach I've described here. 😄